numpy/_core/src/umath/loops_trigonometric.dispatch.cpp

```diff
@@ -0,0 +1,298 @@
+    //npy_intp load_index[lanes/2];
+    //for (auto i = 0; i < lanes; ++i) {
+    //    load_index[i] = i * ssrc;
+    //}
+    //vec_s32 vec_lindex = hn::LoadU(s32, load_index);
+    //npy_intp store_index[lanes/2];
+    //for (auto i = 0; i < lanes; ++i) {
+    //    store_index[i] = i * sdst;
+    //}
+    //vec_s32 vec_sindex = hn::LoadU(s32, store_index);
+
+    for (; len > 0; len -= lanes, src += ssrc*lanes, dst += sdst*lanes) {
+        vec_f32 x_in;
+        if (ssrc == 1) {
+            x_in = hn::LoadN(f32, src, len);
+        } else {
+            x_in = GatherIndexN(src, ssrc, len);
+        }
+        opmask_t nnan_mask = hn::Not(hn::IsNaN(x_in));
+        // Eliminate NaN to avoid FP invalid exception
+        x_in = hn::IfThenElse(nnan_mask, x_in, zerosf);
+        opmask_t simd_mask = hn::Le(hn::Abs(x_in), max_cody);
+        npy_uint64 simd_maski;
+        hn::StoreMaskBits(f32, simd_mask, (uint8_t*)&simd_maski);
+        /*
+         * For elements outside of this range, Cody-Waite's range reduction
+         * becomes inaccurate and we will call libc to compute cosine for
+         * these numbers
+         */
+        if (!hn::AllFalse(f32, simd_mask)) {
+            vec_f32 x = hn::IfThenElse(hn::And(nnan_mask, simd_mask), x_in, zerosf);
+
+            vec_f32 quadrant = hn::Mul(x, two_over_pi);
+            // round to nearest, -0.0f -> +0.0f, and |a| must be <= 0x1.0p+22
+            quadrant = hn::Add(quadrant, rint_cvt_magic);
+            quadrant = hn::Sub(quadrant, rint_cvt_magic);
+
+            // Cody-Waite's range reduction algorithm
+            vec_f32 reduced_x = simd_range_reduction_f32(
+                x, quadrant, codyw_pio2_highf, codyw_pio2_medf, codyw_pio2_lowf
+            );
+            vec_f32 reduced_x2 = hn::Mul(reduced_x, reduced_x);
+
+            // compute cosine and sine
+            vec_f32 cos = simd_cosine_poly_f32(reduced_x2);
+            vec_f32 sin = simd_sine_poly_f32(reduced_x, reduced_x2);
+
+            vec_s32 iquadrant = hn::NearestInt(quadrant);
+            if (trig_op == SIMD_COMPUTE_COS) {
+                iquadrant = hn::Add(iquadrant, ones);
+            }
+            // blend sin and cos based on the quadrant
+            opmask_t sine_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, ones), hn::Zero(s32)));
+            cos = hn::IfThenElse(sine_mask, sin, cos);
+
+            // multiply by -1 for appropriate elements
+            opmask_t negate_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, twos), twos));
+            cos = hn::MaskedSubOr(cos, negate_mask, zerosf, cos);
+            cos = hn::IfThenElse(nnan_mask, cos, hn::Set(f32, NPY_NANF));
+
+            if (sdst == 1) {
+                hn::StoreN(cos, f32, dst, len);
+            } else {
+                ScatterIndexN(cos, dst, sdst, len);
+            }
+        }
+        if (simd_maski != (npy_uint64)((1 << lanes) - 1)) {
+            float ip_fback[hn::Lanes(f32)];
```

Comment:
Unsure if the benchmarks hit this case often, but it'd be worth checking this compiles to a vector without the alignment attributes (I *think* the Highway thing here would be `hwy::AllocateAligned`)